consensus/hotstuff/cruisectl/block_rate_controller.go

```diff
@@ -0,0 +1,160 @@
+// moving average (EWMA) block rates, computes the target block rate,
+// and computes the error terms.
+type measurement struct {
+	view            uint64    // v       - the current view
+	time            time.Time // t[v]    - when we entered view v
+	blockRate       float64   // r[v]    - measured instantaneous block rate at view v
+	aveBlockRate    float64   // r_N[v]  - EWMA block rate over past views [v-N, v]
+	targetBlockRate float64   // r_SP[v] - computed target block rate at view v
+	proportionalErr float64   // e_N[v]  - proportional error at view v
+	integralErr     float64   // E_N[v]  - integral of error at view v
+	derivativeErr   float64   // ∆_N[v]  - derivative of error at view v
+}
+
+// epochInfo stores data about the current and next epoch. It is updated when we enter
+// the first view of a new epoch, or the EpochSetup phase of the current epoch.
+type epochInfo struct {
+	curEpochFinalView        uint64
+	curEpochTargetSwitchover time.Time
+	nextEpochFinalView       *uint64
+	epochFallbackTriggered   *atomic.Bool
+}
+
+// BlockRateController dynamically adjusts the block rate delay of this node,
+// based on the measured block rate of the consensus committee as a whole, in
+// order to achieve a target overall block rate.
+type BlockRateController struct {
+	component.Component
+
+	config *Config
+	state  protocol.State
+	log    zerolog.Logger
+
+	lastMeasurement *measurement    // the most recently taken measurement
+	blockRateDelay  *atomic.Float64 // the block rate delay value to use when proposing a block
+	epochInfo
+
+	viewChanges chan uint64       // OnViewChange events           (view entered)
+	epochSetups chan *flow.Header // EpochSetupPhaseStarted events (block header within setup phase)
+}
+
+// NewBlockRateController returns a new BlockRateController.
+func NewBlockRateController(log zerolog.Logger, config *Config, state protocol.State) (*BlockRateController, error) {
+
+	ctl := &BlockRateController{
+		config:      config,
+		log:         log,
+		state:       state,
+		viewChanges: make(chan uint64, 1),
+		epochSetups: make(chan *flow.Header, 1),
+	}
+
+	ctl.Component = component.NewComponentManagerBuilder().
+		AddWorker(ctl.processEventsWorker).
+		Build()
+
+	// TODO initialize last measurement
+	// TODO initialize epoch info
+	_ = ctl.lastMeasurement
+	_ = ctl.curEpochTargetSwitchover
+	_ = ctl.curEpochFinalView
+	_ = ctl.nextEpochFinalView
+
+	return ctl, nil
+}
+
+// BlockRateDelay returns the current block rate delay value to use when proposing, in milliseconds.
+// This function reflects the most recently computed output of the PID controller
+func (ctl *BlockRateController) BlockRateDelay() float64 {
+	return ctl.blockRateDelay.Load()
+}
+
+// processEventsWorker is a worker routine which processes events received from other components.
+func (ctl *BlockRateController) processEventsWorker(ctx irrecoverable.SignalerContext, ready component.ReadyFunc) {
```

Comment:
I would challenge the correctness of the goDoc, because this is _not_ a worker routine. I prefer to clearly distinguish between the worker (in my mind that is largely a go routine) from the task / worker logic.


```suggestion
// processEventsWorker is the logic for processing events received from other components. 
// This method should be executed by a dedicated worker routine (not concurrency safe).
func (ctl *BlockRateController) processEventsWorkerLogic(ctx irrecoverable.SignalerContext, ready component.ReadyFunc) {
```

Reply:
I will change this while registering my [disagreement](https://github.com/onflow/flow-go/pull/4187#discussion_r1190330059). 😉

Reply:
https://github.com/onflow/flow-go/pull/4332/commits/c8a1665d9c8b39b1a142011c22da331c766e3c71